src/test/java/io/reactivex/observers/TestObserverTest.java

```diff
@@ -1501,4 +1501,128 @@ public void assertValuesOnlyThrowsWhenErrored() {
             // expected
         }
     }
+
+    @Test
+    public void assertValueSetOnly() {
+        TestObserver<Integer> to = TestObserver.create();
+        to.onSubscribe(Disposables.empty());
+        to.assertValueSetOnly(Collections.<Integer>emptySet());
+
+        to.onNext(5);
+        to.assertValueSetOnly(Collections.singleton(5));
+
+        to.onNext(-1);
+        to.assertValueSetOnly(new HashSet<Integer>(Arrays.asList(5, -1)));
+    }
+
+    @Test
+    public void assertValueSetOnlyThrowsOnUnexpectedValue() {
+        TestObserver<Integer> to = TestObserver.create();
+        to.onSubscribe(Disposables.empty());
+        to.assertValueSetOnly(Collections.<Integer>emptySet());
+
+        to.onNext(5);
+        to.assertValueSetOnly(Collections.singleton(5));
+
+        to.onNext(-1);
+
+        try {
+            to.assertValueSetOnly(Collections.singleton(5));
+            fail();
```

Comment:
`fail` throws `AssertionError` so this test always passes. Perhaps throw a `RuntimeException` or add a fail message and verify the `catch` wasn't that message. Also this is overlooked in many other test using `fail` and `catch (AssertionError)` so it would be great if this could be fixed as well.